index.js

```diff
@@ -1,39 +1,224 @@
-function CoreStore (storage) {
-  if (!(this instanceof CoreStore)) return new CoreStore(storage)
+var hypercore = require('hypercore')
+var hcrypto = require('hypercore-crypto')
+var deferred = require('deferred-random-access')
+var TinyBox = require('tinybox')
+var path = require('path')
+var LRU = require('lru')
+var { nextTick } = process
+var DKEY = 'd!' // hex discovery key  to key
+var LKEY = 'l!' // hex key to local name
+var INV_LKEY = 'L!' // local name to key
+var FEED = 'f_'
+
+function Storage (storage, opts) {
+  if (!(this instanceof Storage)) return new Storage(storage, opts)
+  if (!opts) opts = {}
+  this._db = new TinyBox(storage('db'))
+  this._storage = storage
+  this._feeds = {} // map hypercore keys to feeds
+  this._dkeys = {} // map discovery key to key for loaded feeds
+  this._lnames = {} // map local names to key for loaded feeds
+  this._delete = opts.delete
+}
+
+Storage.prototype._storageF = function (prefix) {
+  var self = this
+  return function (p) {
+    return self._storage(path.join(prefix, p))
+  }
 }
 
 // Map a discovery key to a public key.
-CoreStore.prototype.fromDiscoveryKey = function (discoKey, cb) {
+Storage.prototype.fromDiscoveryKey = function (dkey, cb) {
+  var hdkey = asHexStr(dkey)
+  if (this._dkeys.hasOwnProperty(hdkey)) {
+    return nextTick(cb, null, this._dkeys[hdkey])
+  }
+  this._db.get(DKEY + hdkey, function (err, node) {
+    if (err) cb(err)
+    else cb(null, node === null ? null : node.value)
```

Comment:
For a small perf increase, this codepath could set `self._dkeys[hdkey] = node.value`